File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.st = []
        self.mini = []
        

    def push(self, val: int) -> None:
        self.st.append(val)
        if not self.mini or val <= self.mini[-1]:
            self.mini.append(val)
        

    def pop(self) -> None:
        val = self.st.pop()
        if val == self.mini[-1]:
            self.mini.pop()

    def top(self) -> int:
        return self.st[-1]
        

    def getMin(self) -> int:
        return self.mini[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 155-min-stack/min-stack.cpp
class MinStack {
public:
    stack<int> st;
    stack<int> minStack;

    MinStack() {}

    void push(int val) {
        st.push(val);
        if (minStack.empty() || val <= minStack.top()) {
            minStack.push(val);
        }
    }

    void pop() {
        int val = st.top();
        if (val == minStack.top()) {
            minStack.pop();
        }
        st.pop();
        // return val;
    }

    int top() { return st.top(); }

    int getMin() {
        if (minStack.empty())
            return -1;
        return minStack.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */